Improve setup for ALICA logging

bjoernschroeder opened the pull request:
Wrike: https://www.wrike.com/open.htm?id=905279346

Introduces a new logging. The logger can be implemented by the developer and set, like communication. An interface IAlicaLogger is provided.

Updates the old logging ALICA_VERBOSITY_MSG with new logging. To set the verbosity level of the logger, a verbosity level can be set in the Alica.yaml config file.

Commit:
add IAlicaLogger, add comment to logger

Commit:
add IAlicaLogger, add setter for logger in context, initialize logger when not set

Commit:
replace old logging with new logging

Commit:
add TestLogger

Commit:
create fixture with logging

Commit:
update use of query

Commit:
add test for logger, set log verbosity in config

Commit:
update AlicaRosLogger implementation

Commit:
Merge branch 'rr_devel' into bas_logger

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -605,7 +664,12 @@ bool AlicaContext::setOptions(const std::vector<std::pair<std::string, T>>& keyV
     } catch (const YAML::Exception& e) {
-        ALICA_WARNING_MSG("AC: Could not set config values: " << e.msg);
+        if (!_logger) {
+            ALICA_WARNING_MSG("AC: Could not set config values: " << e.msg);
```
can use std::cerr & get rid of these macros

veerajsk commented on alica_engine/include/engine/IAlicaLogger.h:
```diff
@@ -0,0 +1,39 @@
+class IAlicaLogger
+{
+public:
+    IAlicaLogger() = default;
```
constructor definition is not required for an interface

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -558,6 +583,35 @@ void AlicaContext::setWorldModel(Args&&... args)
+        return;
+    }
+
+    // TODO: Read verbosity from config, pass to logger
```
why the TODO here?
can pass the agent name as well to the constructor

veerajsk commented on alica_engine/include/engine/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,39 @@
+        }
+
+        if (verbosity == alica::Verbosity::DEBUG || verbosity == alica::Verbosity::INFO) {
+            std::cout << msg << std::endl;
```
can prepend the agent name to the msg

veerajsk commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -81,10 +82,7 @@ class RunnableObject
-    {
-        _runnableObjectTracer.setTracing(type, customTraceContextGetter);
-    }
+    void setTracing(TracingType type, std::function<std::optional<std::string>()> customTraceContextGetter = {});
```
this is a bit complicated. Instead, can set the logger in the constructor & store it in the TraceRunnableObject as a protected member. We want to remove engine references in all modules as well

veerajsk commented on alica_engine/src/engine/teammanager/TeamManager.cpp:
```diff
@@ -103,7 +105,7 @@ void TeamManager::readSelfFromConfig(const YAML::Node& config)
         } else {
             _localAnnouncement.senderID = _engine->generateID();
-            ALICA_DEBUG_MSG("TM: Auto generated id " << _localAnnouncement.senderID);
+            _engine->getLogger().log(Verbosity::DEBUG, "TM: Auto generated id ", _localAnnouncement.senderID);
```
use the logger in the team manager & not from the engine

veerajsk commented on alica_engine/src/engine/Logger.cpp:
```diff
@@ -79,7 +76,7 @@ void Logger::processString(const std::string& event)
     }
     _receivedEvent = true;
-    ALICA_DEBUG_MSG("Logger: " << _eventStrings.back());
+    _ae->getLogger().log(Verbosity::DEBUG, "Logger: ", _eventStrings.back());
```
remove engine reference, store the logger in the class instead.

veerajsk reviewed (commented):
there are places where the logger is accessed via the engine. We plan to remove engine references throughtout alica components. So instead store the logger as a local reference variable in these modules & access via that instead

Commit:
remove ALICA_DEBUG_MSG macro use, remove access of logger via engine, pass logger to objects via constructor

Commit:
pass logger to RunnableObject, remove use of engine for accessing logger, remove debug include from utilityFunction

Commit:
add logging verbosity to config, update query creation for supplementary tests

Commit:
add logger to utility functions

Commit:
update codegen for using logger with utility functions

Commit:
format

Commit:
Merge branch 'rr_devel' into bas_logger

veerajsk commented on alica_engine/include/engine/IAlicaLogger.h:
```diff
@@ -0,0 +1,39 @@
+    virtual ~IAlicaLogger() = default;
+
+    template <class... Args>
+    void log(Verbosity verbosity, Args&&... args)
```
this is more of a convenience function & need not be part of the interface. Further, since it's named the same as the virtual function it can get confusing especially when inheretance & templates get involved. I think this function can be taken out of the class & implemented as a free function under some namespace. Eg.
```
namespace Logging {
template <class... Args>
void log(IAlicaLogger& logger, Verbosity verbosity, Args&&... args) {
    ...
    logger.log(verbosity, logStr);
}
};
```

triccyx replied:
I don't know exactly the contest. But have you consider also to add:
an `operator<<` for logging? It could be intuitive and comfortable.
You can have 
operator<<(Verbosity xx) //for verbosity
 operator<<(endl_type endl) //for endl
and another for other types 
This is just a suggestion

Dale-Koenig replied:
I would also request a method returning ostream that we can use with << like in Ros
I suppose it might need to print out \n when the ostream destructs

Would also be nice to have convenience functions for different verbosity, like logWarn

veerajsk commented on alica_engine/include/engine/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,41 @@
+    {
+    }
+
+    virtual ~AlicaDefaultLogger() = default;
```
need not be virtual

triccyx replied:
I think you can remove

veerajsk commented on alica_engine/include/engine/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,41 @@
+
+    virtual ~AlicaDefaultLogger() = default;
+
+    void log(const std::string& msg, Verbosity verbosity)
```
use `override` to convey the fact that this method overrides its base class counterpart

veerajsk commented on alica_engine/include/engine/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,41 @@
+            return;
+        }
+
+        if (verbosity == alica::Verbosity::DEBUG || verbosity == alica::Verbosity::INFO) {
```
can use `<` to simplify i.e.
```
if (verbosity < alica::Verbosity::WARNING) {
    std::cout
} else {
    std::cerr
}
```

veerajsk replied:
1. Print the verbosity level in the log
2. nit: would be nice to print colored output: https://www.codegrepper.com/code-examples/cpp/c%2B%2B+cout+with+color

veerajsk commented on alica_engine/include/engine/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,41 @@
+    void log(const std::string& msg, Verbosity verbosity)
+    {
+        // dont print log message when logger verbosity is greater than log verbosity.
+        if (_verbosity > verbosity) {
```
move the implementation to cpp file

veerajsk commented on supplementary/alica_ros_proxy/include/logger/AlicaRosLogger.h:
```diff
@@ -0,0 +1,58 @@
+
+    virtual ~AlicaRosLogger() = default;
+
+    template <class... Args>
```
why is this a template? Isn't it overriding the base class's log method? This is why you should use `override` & also not overload virtual methods in the base class (that too a template overrload)

veerajsk commented on supplementary/alica_ros_proxy/include/logger/AlicaRosLogger.h:
```diff
@@ -0,0 +1,58 @@
+    void log(std::string& msg, Verbosity verbosity)
+    {
+        if (verbosity == alica::Verbosity::DEBUG) {
+            ROS_DEBUG("%s", msg.c_str());
```
can use ROS_*_STREAM_NAMED_*_ methods & use the agent name as the name (taken via the constructor), just like the default logger.

veerajsk commented on supplementary/alica_ros_proxy/include/logger/AlicaRosLogger.h:
```diff
@@ -0,0 +1,58 @@
+public:
+    AlicaRosLogger(Verbosity verbosity)
+    {
+        ros::console::Level level;
```
implementation should be in the cpp file

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -411,6 +421,18 @@ class AlicaContext
+     * @note LoggerType must be a derived class of IAlicaLogger
+     * @note This must be called before initializing context
+     *
+     * @param args Arguments to be forwarded to constructor of logger. Might be empty.
```
you should also mention that the first 2 arguments given to the constructor will be the `Verbosity` & `AgentName`

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -561,6 +584,34 @@ void AlicaContext::setWorldModel(Args&&... args)
+        verbosity = Verbosity::INFO;
+    } else if (verbosityString == "ERROR") {
+        verbosity = Verbosity::ERROR;
+    } else if (verbosityString == "FATAL") {
```
`WARNING` is missing

veerajsk commented on alica_engine/include/engine/constraintmodul/VariableSyncModule.h:
```diff
@@ -109,8 +109,8 @@ int VariableSyncModule::getSeeds(const std::vector<VarType*>& query, const std::
 #ifdef RS_DEBUG
-    std::cout << "RS: Generated " << seeds.size() << "seeds" << std::endl;
-    for (int i = 0; i < seeds.size(); ++i) {
+    _ae->getLogger().log(Verbosity::DEBUG, "RS: Generated ", seeds.size(), "seeds") for (int i = 0; i < seeds.size(); ++i)
```
remove engine reference. Will need to store logger in this class

veerajsk commented on alica_engine/include/engine/AlicaEngine.h:
```diff
@@ -102,6 +102,7 @@ class AlicaEngine
     const IAlicaCommunication& getCommunicator() const;
     const AlicaClock& getAlicaClock() const;
     IAlicaTimerFactory& getTimerFactory() const;
+    IAlicaLogger& getLogger() const;
```
don't need this. The context will provide the API to get the logger to the user & alica internals should not access the logger via the engine

veerajsk commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -81,27 +82,27 @@ class RunnableObject
     RunningPlan* getPlanContext() const { return _runningplanContext; }
     const std::shared_ptr<Blackboard> getBlackboard() { return _blackboard; }
     IAlicaWorldModel* getWorldModel() { return _wm; };
+    IAlicaLogger& getLogger() { return _logger; };
```
not sure if this is required. The application need not use alica's logger through this interface. It can anyways specify its own logger at init time to the engine

veerajsk commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -15,9 +15,9 @@ void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb,
         try {
             childBb.set(childKey, lockedParentBb.get(parentKey));
-            ALICA_DEBUG_MSG("passing " << parentKey << " into " << childKey);
+            std::cout << "passing " << parentKey << " into " << childKey << std::endl;
```
this should use logger. Can take as a parameter

veerajsk commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -68,7 +66,8 @@ void ConditionStore::addCondition(const Condition* con)
     }
 
-    ALICA_DEBUG_MSG("CS: Added condition in " << con->getAbstractPlan()->getName() << " with " << con->getVariables().size() << " variables. CS: " << this);
+    std::cout << "CS: Added condition in " << con->getAbstractPlan()->getName() << " with " << con->getVariables().size() << " variables. CS: " << this
```
should use logger

veerajsk commented on alica_engine/src/engine/modelmanagement/factories/Factory.cpp:
```diff
@@ -160,7 +159,7 @@ void Factory::storeElement(AlicaElement* ael, const std::string& type)
     } else if (alica::Strings::variableBinding.compare(type) == 0) {
         // case for ignored types
-        ALICA_DEBUG_MSG("Factory: INFO: Element type " << type << " is not stored in plan repository.");
+        std::cout << "Factory: INFO: Element type " << type << " is not stored in plan repository." << std::endl;
```
should use logger

veerajsk commented on alica_engine/src/engine/Logger.cpp:
```diff
@@ -17,23 +17,21 @@
         , _receivedEvent(false)
         , _inIteration(false)
         , _logging(false)
+        , _logger(logger)
```
should rename either this variable or this class. Having a logger member variable in a Logger class is weird

veerajsk commented on alica_engine/src/engine/UtilityFunction.cpp:
```diff
@@ -168,22 +168,23 @@ UtilityInterval UtilityFunction::getPriorityResult(IAssignment ass) const
 
-    ALICA_DEBUG_MSG("##" << std::endl << "UF: prioUI = " << priResult);
-    ALICA_DEBUG_MSG("UF: denum = " << denum);
+    // TODO: fix
```
what needs to be fixed here?

veerajsk reviewed (commented):
Looking at all the places the logger needs to be stored, accessed & passed around, I think it is valid case to use a singleton instead, which will be created on init of the engine & destroyed on terminate of the engine (see the implementation of Lbc singleton).

triccyx commented on alica_engine/include/engine/constraintmodul/Query.h:
```diff
@@ -136,7 +138,7 @@ bool Query::getSolution(const RunningPlan* pi, std::vector<ResultType>& result)
 
     if (!pi->getAlicaEngine()->existSolver<SolverType>()) {
-        std::cerr << "Query::getSolution: The engine does not have a suitable solver for the given type available." << std::endl;
+        _logger.log(Verbosity::ERROR, "Query::getSolution: The engine does not have a suitable solver for the given type available.");
```
Example as suggested before:
_logger<<Verbosity::ERROR<<"Query::getSolution: The engine does not have a suitable solver for the given type available."
Some work on Verbosity need todo
(only a suggestion)

triccyx commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -15,9 +15,9 @@ void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb,
+            std::cout << "passing " << parentKey << " into " << childKey << std::endl;
         } catch (std::exception& e) {
-            ALICA_WARNING_MSG("Blackboard error passing " << parentKey << " into " << childKey << ". " << e.what());
+            std::cerr << "Blackboard error passing " << parentKey << " into " << childKey << ". " << e.what() << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -29,9 +29,9 @@ void BlackboardUtil::setOutput(Blackboard* parent_bb, const Blackboard* child_bb
         try {
             lockedParentBb.set(parentKey, childBb.get(childKey));
-            ALICA_DEBUG_MSG("passing " << childKey << " into " << parentKey);
+            std::cout << "passing " << childKey << " into " << parentKey << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -29,9 +29,9 @@ void BlackboardUtil::setOutput(Blackboard* parent_bb, const Blackboard* child_bb
+            std::cout << "passing " << childKey << " into " << parentKey << std::endl;
         } catch (std::exception& e) {
-            ALICA_WARNING_MSG("Blackboard error passing " << childKey << " into " << parentKey << ". " << e.what());
+            std::cerr << "Blackboard error passing " << childKey << " into " << parentKey << ". " << e.what() << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -102,15 +101,15 @@ void ConditionStore::removeCondition(const Condition* con)
     }
 
-    ALICA_DEBUG_MSG("CS: Removed condition in " << con->getAbstractPlan()->getName() << " with " << con->getVariables().size() << " variables.");
+    std::cout << "CS: Removed condition in " << con->getAbstractPlan()->getName() << " with " << con->getVariables().size() << " variables." << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -102,15 +101,15 @@ void ConditionStore::removeCondition(const Condition* con)
 void ConditionStore::acceptQuery(Query& query, const RunningPlan* rp) const
 {
-    ALICA_DEBUG_MSG("ConditionStore: Accepting Query - Active conditions in store is " << _activeConditions.size() << " CS: " << this);
+    std::cout << "ConditionStore: Accepting Query - Active conditions in store is " << _activeConditions.size() << " CS: " << this << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -121,7 +120,7 @@ void ConditionStore::acceptQuery(Query& query, const RunningPlan* rp) const
     }
 
-    ALICA_DEBUG_MSG("ConditionStore: Query contains static variables: " << staticVarBuffer.getCurrent());
+    std::cout << "ConditionStore: Query contains static variables: " << staticVarBuffer.getCurrent() << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -133,16 +132,16 @@ void ConditionStore::acceptQuery(Query& query, const RunningPlan* rp) const
             staticVarBuffer.editNext().push_back(curStaticVariable);
 
-            ALICA_DEBUG_MSG("ConditionStore: Checking static variable: " << *curStaticVariable);
+            std::cout << "ConditionStore: Checking static variable: " << *curStaticVariable << std::endl;
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -133,16 +132,16 @@ void ConditionStore::acceptQuery(Query& query, const RunningPlan* rp) const
 
-            ALICA_DEBUG_MSG(
-                    "ConditionStore: Conditions active under variable " << *activeVar2CondMapEntry->first << ": " << activeVar2CondMapEntry->second.size());
+            std::cout << "ConditionStore: Conditions active under variable " << *activeVar2CondMapEntry->first << ": " << activeVar2CondMapEntry->second.size()
```
Could you pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/model/Condition.cpp:
```diff
@@ -32,15 +30,15 @@ void Condition::setConditionString(const std::string& conditionString)
 {
     if (_basicCondition == nullptr) {
-        ALICA_ERROR_MSG("Condition: Missing implementation of condition: ID " << getId());
+        std::cerr << "Condition: Missing implementation of condition: ID " << getId() << std::endl;
```
Is it possible  to pass also the logger to this class/method?

triccyx commented on alica_engine/src/engine/model/Condition.cpp:
```diff
@@ -32,15 +30,15 @@ void Condition::setConditionString(const std::string& conditionString)
             ret = _basicCondition->evaluate(const_cast<RunningPlan&>(rp).getSharedPointer(), wm);
         } catch (const std::exception& e) {
-            ALICA_ERROR_MSG("Condition: Exception during evaluation catched: " << std::endl << e.what());
+            std::cerr << "Condition: Exception during evaluation catched: " << std::endl << e.what() << std::endl;
```
Is it possible  to pass also the logger to this class/method?

triccyx commented on alica_tests/src/test/test_alica_logger.cpp:
```diff
@@ -0,0 +1,90 @@
+    ASSERT_EQ(logger.logs.at(14).first, Verbosity::DEBUG);
+}
+} // namespace
+} // namespace alica
```
In general, is better to write the check in this way:
ASSERT_NE(expected,value) instead of ASSERT_TRUE(...)
because the test log clearly indicates the expected on the left and the real value on the right, in the first case.
Just a suggestion for future use, not necessary to change

Commit:
Merge branch 'rr_devel' into bas_logger

Dale-Koenig commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -608,7 +664,12 @@ bool AlicaContext::setOptions(const std::vector<std::pair<std::string, T>>& keyV
     } catch (const YAML::Exception& e) {
-        ALICA_WARNING_MSG("AC: Could not set config values: " << e.msg);
+        if (!_logger) {
+            std::cerr << "AC: Could not set config values: " << e.msg << std::endl;
```
Can we not use the default logger for this?

bjoernschroeder replied:
We can create a local temporary instance of a default logger to log these msgs, but we cant initialize the default logger with setLogger, because that sets the DefaultLogger as a singleton.

veerajsk replied:
I think its fine for it be `std::cerr` since the logger is not set. Initializing the default logger here could be tricky & probably not worth it

veerajsk replied:
maybe you can additionally print `Logging to console since logger is not initialized` in the log msg

Commit:
move jars to LFS

Commit:
- Change logger into singleton
- prevent singleton destruction
- add verbosity level warning when initializing
- remove logger from constructors
- move logger files into logging

Commit:
remove logger from generated code

Commit:
update AlicaRosLogger

Commit:
replace prints with logs, format

Commit:
remove unused destructor

Commit:
- support logging with << operator
- update logging to use <<
- add logFunctions for each verbosity level
- remove unused includes & forward declarations
- revert log in Logging.h to old format

Commit:
format

Commit:
update codegen jars

veerajsk commented on alica_engine/include/engine/logging/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,23 @@
+    void log(const std::string& msg, Verbosity verbosity) override;
+
+private:
+    Verbosity _verbosity;
```
`const Verbosity _verbosity`

veerajsk commented on alica_engine/include/engine/logging/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,23 @@
+
+private:
+    Verbosity _verbosity;
+    std::string _localAgentName;
```
`const std::string _localAgentName`

veerajsk commented on alica_engine/include/engine/logging/AlicaDefaultLogger.h:
```diff
@@ -0,0 +1,23 @@
+private:
+    Verbosity _verbosity;
+    std::string _localAgentName;
+    std::unordered_map<alica::Verbosity, std::pair<std::string, std::string>> _verbosityMap;
```
can make map static & const
can use a better variable name
can store strings as const char*
`static const std::unordered_map<alica::Verbosity, std::pair<const char*, const char*>> _verbosityStringAndColorMap`

Dale-Koenig replied:
Add comment to clarify what this is

veerajsk commented on alica_engine/src/engine/logging/AlicaDefaultLogger.cpp:
```diff
@@ -0,0 +1,42 @@
+
+#include <iostream>
+
+#define RESET "\033[0m"
```
better to define them as `static constexpr const char* RESET = "\033[0m"` inside the class `AlicaDefaultLogger`. Macro names could collide & cause problems & in general should be avoided as much as possible

veerajsk commented on alica_engine/include/engine/logging/LoggingUtil.h:
```diff
@@ -0,0 +1,47 @@
+{
+public:
+    LogEntry(Verbosity v);
+    LogEntry(LogEntry&& entry);
```
can be defaulted I think

veerajsk replied:
even better, can delete the move constructor since we don't want to copy/move objects of this class anyway

veerajsk commented on alica_engine/include/engine/logging/LoggingUtil.h:
```diff
@@ -0,0 +1,47 @@
+#include "engine/logging/AlicaLogger.h"
+namespace alica
+{
+namespace Logging
```
namespace `Logging` is not required since the logging methods are anyway accessed via the `LoggingUtil` class. Prevents having to type `Logging::LoggingUtil`. Maybe also rename `LoggingUtil` to `Logging`

veerajsk reviewed (commented):
reviewed everything under the logging folder, will complete the rest tomorrow

Dale-Koenig commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -533,7 +546,7 @@ template <class TraceFactoryType, class... Args>
 {
     if (_initialized) {
-        ALICA_WARNING_MSG("AC: Context already initialized. Can not set new tracefactory");
+        Logging::LoggingUtil::logWarn() << "AC: Context already initialized. Can not set new tracefactory";
```
Can we take the log "namespace" (in this case the `AC` part) and have it as an optional argument to `logWarn()`?
Then we can later configure things like are available in ROS logger:
- Filter logs of certain namespaces based on yaml config
- Configure displaying/not displaying the namespace
- Configure order and show namespace e.g. before the timestamp

veerajsk replied:
can do, but will require adding a `std::string namespace` argument to the `log` interface in `IAlicaLogger`

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -411,6 +411,19 @@ class AlicaContext
+     * The first two arguments, verbosity and agentName, will always be passed to the constructor of the logger.
+     */
+    template <class LoggerType, class... Args>
+    void setLogger(Args&&... args);
```
let's make this method static since we are using a singleton for the Logger. This would mean that we will only be able to create 1 instance per process

veerajsk replied:
can also mention this 1 logger per process thing in the comment

bjoernschroeder replied:
I think this is tricky, because we use the member variables _localAgentName, which can be passed to the static setLogger function, and _configRootNode. For the config, we would have to pass the required information to parse the config to setLogger, parse the config to then read the verbosity level. Then we have to parse the config again in the initializer list of the AlicaContext.

veerajsk replied:
you could take the alica context as a parameter, but that may make the API a bit confusing, so lets leave it as is, but mention in a comment that logger will be static & limited to 1 instance per process

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -561,6 +574,31 @@ void AlicaContext::setWorldModel(Args&&... args)
+        verbosity = Verbosity::ERROR;
+    } else if (verbosityString == "FATAL") {
+        verbosity = Verbosity::FATAL;
+    }
```
else, set verbosity to `INFO` or just initialize verbosity to `INFO`
nit: can use a static const unordered_map to map from verbosity to string instead of an if else ladder -> easier to read

veerajsk commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -44,6 +44,7 @@ AlicaEngine::AlicaEngine(AlicaContext& ctx, YAML::Node& config, const AlicaConte
         , _stepEngine(alicaContextParams.stepEngine)
         , _modelManager(_configChangeListener, alicaContextParams.configPath, _planRepository)
         , _masterPlan(_modelManager.loadPlanTree(alicaContextParams.masterPlanName))
+        , _planRepository()
```
can you place this before `_modelManager`? That's the declaration order in the .h file & that's the order in which initialization takes place.

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -212,11 +212,11 @@ class AlicaContext
      * @return Return code '0' stands for success, any other for corresponding error
      */
-    int terminate();
+    int terminate(bool preventSingletonDestruction = false);
```
hmm, looks like we should always `not` destroy the logger to keep the API simple -> can remove this parameter in that case

veerajsk commented on alica_engine/src/engine/constraintmodul/ConditionStore.cpp:
```diff
@@ -68,7 +68,8 @@ void ConditionStore::addCondition(const Condition* con)
     }
 
-    ALICA_DEBUG_MSG("CS: Added condition in " << con->getAbstractPlan()->getName() << " with " << con->getVariables().size() << " variables. CS: " << this);
+    Logging::LoggingUtil::log(
```
nit: use `logDebug` like in the rest of the code base

veerajsk commented on alica_engine/src/engine/syncmodule/SynchronisationProcess.cpp:
```diff
@@ -5,14 +5,12 @@
 #include "engine/syncmodule/SyncRow.h"
 
-//#define ALICA_DEBUG_LEVEL_DEBUG
-#include <alica_common_config/debug_output.h>
```
can this file be removed completely?

veerajsk commented on alica_tests/src/test/Expr/src/Behaviour/ConstraintUsingBehaviour.cpp:
```diff
@@ -37,16 +37,17 @@ void ConstraintUsingBehaviour::run(void* msg)
     _callCounter = 0;
-    _query.clearStaticVariables();
-    _query.addStaticVariable(getVariable("Y"));
+    _query = std::make_unique<Query>(getLogger());
```
fix: no need to take in the logger

bjoernschroeder replied:
fixed in https://github.com/rapyuta-robotics/alica/pull/302

veerajsk commented on alica_tests/src/test/etc/nase/Alica.yaml:
```diff
@@ -62,6 +62,9 @@ Alica:
     LogFolder: /tmp/log
 
+  Logging:
+    Verbosity: DEBUG
```
use `INFO`, debug spams at the moment

bjoernschroeder replied:
changed the level to INFO in https://github.com/rapyuta-robotics/alica/pull/302

veerajsk commented on alica_tests/src/test/test_task_assignment.cpp:
```diff
@@ -82,7 +83,8 @@ TEST_F(TaskAssignmentTest, constructTaskAssignment)
     // fake inform the team observer about roles of none existing robots
-    alica::RunningPlan* rp = new RunningPlan(ae, ae->getPlanRepository().getPlans().find(1407152758497), nullptr);
+    alica::AlicaDefaultLogger logger(alica::Verbosity::DEBUG, "nase");
+    alica::RunningPlan* rp = new RunningPlan(ae, ae->getPlanRepository().getPlans().find(1407152758497), nullptr, logger);
```
logger argument not required -> I think you need to fix the tests?

bjoernschroeder replied:
all test changes are now in https://github.com/rapyuta-robotics/alica/pull/302, the logger is not passed to the RP anymore

veerajsk commented on supplementary/alica_codegeneration/alica-plan-designer-fx-codegeneration/src/main/java/de/unikassel/vs/alica/generator/cpp/XtendTemplates.xtend:
```diff
@@ -316,14 +316,9 @@ namespace alica
-    UtilityFunctionCreator::UtilityFunctionCreator()
-    {
-    }
+    UtilityFunctionCreator::~UtilityFunctionCreator() {};
```
changes won't be required to the code generation now right?

veerajsk commented on supplementary/alica_ros_proxy/include/logger/AlicaRosLogger.h:
```diff
@@ -0,0 +1,23 @@
+    void log(const std::string& msg, Verbosity verbosity) override;
+
+private:
+    std::string _localAgentName;
```
const

veerajsk commented on supplementary/alica_ros_proxy/include/logger/AlicaRosLogger.h:
```diff
@@ -0,0 +1,23 @@
+
+private:
+    std::string _localAgentName;
+    int64_t _localAgentId;
```
const alica::AgentId

veerajsk commented on supplementary/alica_ros_proxy/src/logger/AlicaRosLogger.cpp:
```diff
@@ -0,0 +1,56 @@
+        : _localAgentName(localAgentName)
+        , _localAgentId(localAgentId)
+{
+    ros::console::Level level;
```
`level = INFO`
nit: define static const unordered_map & initialize in the initializer list itself

Commit:
fix review comments

Dale-Koenig reviewed (approved):
High level okay

Commit:
remove test changes

Commit:
Merge branch 'rr_devel' into bas_logger

Commit:
remove newline after logSpace, check if logging config value exists before parsing

Commit:
remove unused codegen changes

Commit:
add comment to setLogger

Commit:
remove empty line from codegen

bjoernschroeder committed:
Logger: alica_tests, supplementary_tests and turtlesim (#302)

* regenerate files

* update supplementary tests, move generated files to src/test/Expr

* update turtlesim

* format

* fix review comments

* update turtlesim, supplementary and alica_tests

* regenerate supplementary_tests, delay logger destruction in multiagent tests

* reactivate tests, delay logger destruction

* fix includes

* - remove logspace from tests
- update TestLogger for updated interface
- add log tests to CMake
- fix includes
- add logging values to config

* add logging verbosity to configs, remove delayed singleton deletion, set verbosity to info

* format

* set verbosity to info, increase delay in logger test

* destroy logger after each test, update logger test

* remove print

* regenerate code

* change query to non pointer, remove unused includes

Co-authored-by: bjoernschroeder <bjoernschroder@rapyuta-robotics.com>

bjoernschroeder committed:
Merge branch 'rr_devel' into bas_logger

veerajsk commented on supplementary/alica_ros_turtlesim/alica/Expr/include/constraints/Move1889749086610694100Constraints.h:
```diff
@@ -1,15 +0,0 @@
-class RunningPlan;
-class Constraint1288817888979746811 : public BasicConstraint
-{
-    void getConstraint(std::shared_ptr<ProblemDescriptor> c, std::shared_ptr<RunningPlan> rp);
```
why is this file removed?

bjoernschroeder committed:
move logger initialization to the top of AlicaContext::init

bjoernschroeder committed:
- remove destroy api from logger
- remove TestLogger
- move initialization of logger to the top of AlicaContext::init

bjoernschroeder committed:
format

bjoernschroeder committed:
use SLEEP_UNTIL instead of fix sleep time

bjoernschroeder committed:
use AlicaRosLogger in alica_tests

bjoernschroeder committed:
Merge branch 'rr_devel' into bas_logger

veerajsk commented:
@Dale-Koenig @triccyx @abhi1287 @christianhelm could you please review this PR again, I think there are a few changes since your last review. Also, there is a crash that happens only if the default logger is used (ROS logger works fine) & @bjoernschroeder & me haven't been able to figure out why. This crash only happens in release mode, so there is not much information known about the crash. If you could find anything during the review please comment. PR has been pending since a long time, would be nice to have it merged since it touches a lot of files

triccyx commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -566,6 +579,27 @@ void AlicaContext::setWorldModel(Args&&... args)
+
+    Verbosity verbosity = Verbosity::INFO;
+    // check if config value exists
+    if (_configRootNode["Alica"]["Logging"]["Verbosity"]) {
```
If `Logging` keyword doesn't exist it crashes at the start

triccyx commented on alica_engine/include/engine/logging/Logging.h:
```diff
@@ -0,0 +1,44 @@
+    static void log(const Verbosity verbosity, const std::string& logSpace, Args&&... args)
+    {
+        std::ostringstream oss;
+        (oss << ... << std::forward<Args>(args));
```
If multiple threads call this line? Could be a problem?
Perhaps we need a lock_guard

veerajsk replied:
this line should be fine, each thread will access its own local variable (the function is static)

triccyx commented on alica_engine/src/engine/logging/AlicaDefaultLogger.cpp:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    if (verbosity < alica::Verbosity::WARNING) {
+        std::cout << _verbosityStringAndColorMap.at(verbosity).second << "[" << _verbosityStringAndColorMap.at(verbosity).first << "][" << _localAgentName
```
Could we have problems with multiple threads logging? Lock_guard?

veerajsk replied:
the logs could get interspersed but should not crash I think

veerajsk replied:
https://en.cppreference.com/w/cpp/io/cout: cout is thread safe

Dale-Koenig commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -566,6 +579,27 @@ void AlicaContext::setWorldModel(Args&&... args)
+        std::string verbosityString = _configRootNode["Alica"]["Logging"]["Verbosity"].as<std::string>();
+
+        // check if verbosity level in config is valid
+        if (_verbosityStringToVerbosityMap.find(verbosityString) != _verbosityStringToVerbosityMap.end()) {
```
can do
```suggestion
        if (auto it = _verbosityStringToVerbosityMap.find(verbosityString); it != _verbosityStringToVerbosityMap.end()) {
        verbosity = it->second
```

Dale-Koenig commented on supplementary/alica_ros_proxy/src/logger/AlicaRosLogger.cpp:
```diff
@@ -0,0 +1,47 @@
+    } else if (verbosity == alica::Verbosity::ERROR) {
+        ROS_ERROR_STREAM_NAMED(streamName.str(), msg);
+    } else if (verbosity == alica::Verbosity::FATAL) {
+        ROS_FATAL_STREAM_NAMED(streamName.str(), msg);
```
Not sure if the names here will work unfortunately:  Previously I had trouble with places where the macros would evaluate the name only the first time it was called from a given line of code

Dale-Koenig commented:
> @Dale-Koenig @triccyx @abhi1287 @christianhelm could you please review this PR again, I think there are a few changes since your last review. Also, there is a crash that happens only if the default logger is used (ROS logger works fine) & @bjoernschroeder & me haven't been able to figure out why. This crash only happens in release mode, so there is not much information known about the crash. If you could find anything during the review please comment. PR has been pending since a long time, would be nice to have it merged since it touches a lot of files

Does it occur locally?  Did you try RelWithDebugInfo?

veerajsk commented:
> > @Dale-Koenig @triccyx @abhi1287 @christianhelm could you please review this PR again, I think there are a few changes since your last review. Also, there is a crash that happens only if the default logger is used (ROS logger works fine) & @bjoernschroeder & me haven't been able to figure out why. This crash only happens in release mode, so there is not much information known about the crash. If you could find anything during the review please comment. PR has been pending since a long time, would be nice to have it merged since it touches a lot of files
> 
> Does it occur locally? Did you try RelWithDebugInfo?

@bjoernschroeder

bjoernschroeder commented:
> 
What I have to do to reproduce it locally:
- remove all setLogger calls in alica_tests/include/test_alica.h so we dont use the AlicaRosLogger (+ deactivate test in test_alica_logger.cpp, as it checks if the AlicaRosLogger has been set correctly)
- run stress with 8 cpus
- start tests with catkin run_tests, not with GDB

As soon as visible prints are added to the application, the error might not occure anymore. For example, with additional std::cerr prints even with the DefaultLogger, CI will pass and the error also does not appear locally anymore. However, std::cout prints can be added, because they are not visible in the console by default when running tests with catkin run_tests without GDB. The std::cout prints are later visible in the ros log file after the tests have finished / crashed.

I also tried to catch any exceptions within the DefaultLoggers log method but the tests still crashed without any exception being caught. The error happens in multi agent test, normally when the second agent is initialized in the fixtures setup method, before the actual test is running.

Dale-Koenig commented:
> I also tried to catch any exceptions within the DefaultLoggers log method but the tests still crashed without any exception being caught. The error happens in multi agent test, normally when the second agent is initialized in the fixtures setup method, before the actual test is running.

From this what it sounds like is:
- There is some static initialization that is not thread protected
- You expect the function initializing it to be only called once, but in the test there are two agents on the same PC, so it is called multiple times at approximately the same time
- This causes the crash

If you run in gdb, you can do `r --gtest_catch_exceptions=0` 
Usually if you have exceptions that cause gtest to error out, this will catch them

Dale-Koenig commented on alica_engine/include/engine/logging/Logging.h:
```diff
@@ -0,0 +1,44 @@
+
+    static bool isInitialized();
+};
+} // namespace alica
```
newline at eof

bjoernschroeder commented:
> 

I couldnt reproduce the error in gdb. Tests are always passing in gdb, probably because of the additional prints that gdb does.

veerajsk commented:
> > 
> 
> I couldnt reproduce the error in gdb. Tests are always passing in gdb, probably because of the additional prints that gdb does.

@bjoernschroeder did you try build with RelWithDebugInfo & then run in gdb? This basically builds in release mode but includes debug symbols in the binary.

triccyx commented:
I don't know if it can be helpful and `if it works with ROS` environment but to force the creation of a core dump file
`ulimit -c unlimited`
Then you can analyze it with gdb and some time is possible to find something also in a release compilation.

veerajsk commented:
> I don't know if it can be helpful and `if it works with ROS` environment but to force the creation of a core dump file `ulimit -c unlimited` Then you can analyze it with gdb and some time is possible to find something also in a release compilation.

@bjoernschroeder can refer https://docs.google.com/document/d/1tZV30N7oycmFufgTeSmkV2CbIGNWmoH15ttX5VHTlek/edit if you try to analyze the core dump. Possibly a good option to try

triccyx commented:
If you will be able to create the core 
- The worst case you will have an empty call stack, NOPE with core dump analysis in release
- But usually, in release, you have the call stack without the line number, only the method's name. It could be useful.
- No variable navigation ;(
- It is possible to change the call stack for the various threads

bjoernschroeder committed:
- move communication start after engine init
- dectivate non working tests
- use default logger in alica_tests

bjoernschroeder committed:
Merge branch 'rr_devel' into bas_logger

bjoernschroeder committed:
- use AlicaRosLogger in tests
- move communication start in front of engine init

bjoernschroeder committed:
add new line to logging/Logging.h, use iterator to access _verbosityStringToVerbosity map in AlicaContext
